File: tests/tamm/Test_Mem_Profiler.cpp
#include "ga/macdecls.h"
#include "mpi.h"
#include <chrono>
#include <tamm/tamm.hpp>

using namespace tamm;

using DependencyMap = std::map<IndexVector, TiledIndexSpace>;

void test_tensor_allocate(Scheduler& sch) {
  TiledIndexSpace tis{IndexSpace{range(10)}, 10};
  TiledIndexSpace tis2{IndexSpace{range(20)}, 20};
  TiledIndexSpace tis3{IndexSpace{range(100)}, 20};

  Tensor<double> A{tis, tis, tis, tis};
  Tensor<double> B{tis, tis, tis, tis};
  Tensor<double> C{tis, tis, tis, tis2};
  Tensor<double> D{tis, tis, tis, tis2};
  Tensor<double> E{tis, tis, tis, tis3};

  sch.allocate(A, B, C).deallocate(B, C).execute();

  sch.allocate(D, E).execute();

  print_memory_usage<double>(sch.ec().pg().rank().value());
}

int main(int argc, char* argv[]) {
  tamm::initialize(argc, argv);

  ProcGroup        pg = ProcGroup::create_world_coll();
  ExecutionContext ec{pg, DistributionKind::nw, MemoryManagerKind::ga};

  Scheduler sch{ec};

  test_tensor_allocate(sch);

  tamm::finalize();

  return 0;
}

File: tests/python/Test_Mem_Profiler.py
#!/usr/bin/env python3

import sys
import pytamm as tamm


def test_tensor_allocate(sch):
    tis = tamm.TiledIndexSpace(
        tamm.IndexSpace(tamm.range(10)),
        10,
    )

    tis2 = tamm.TiledIndexSpace(
        tamm.IndexSpace(tamm.range(20)),
        20,
    )

    tis3 = tamm.TiledIndexSpace(
        tamm.IndexSpace(tamm.range(100)),
        20,
    )

    A = tamm.TensorDouble([tis, tis, tis, tis])
    B = tamm.TensorDouble([tis, tis, tis, tis])
    C = tamm.TensorDouble([tis, tis, tis, tis2])
    D = tamm.TensorDouble([tis, tis, tis, tis2])
    E = tamm.TensorDouble([tis, tis, tis, tis3])

    sch.allocate(A, B, C).deallocate(B, C).execute()
    sch.allocate(D, E).execute()

    tamm.print_memory_usage(int(sch.ec().pg().rank()))


def main():
    tamm.initialize(sys.argv, False)

    try:
        pg = tamm.ProcGroup.create_world_coll()

        ec = tamm.ExecutionContext(
            pg,
            tamm.DistributionKind.nw,
            tamm.MemoryManagerKind.ga,
        )

        sch = tamm.Scheduler(ec)

        test_tensor_allocate(sch)

    finally:
        tamm.finalize(True)


if __name__ == "__main__":
    main()